src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -189,6 +198,137 @@ internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projec
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId)
+        {
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            // Iterate over each project referenced by 'projectId', which is now being removed. Update the reverse
+            // references map for the project to no longer include 'projectId' in the list.
+            if (existingForwardReferencesMap.TryGetValue(projectId, out var referencedProjectIds))
+            {
+                foreach (var referencedProjectId in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProjectId, out var reverseDependencies))
+                    {
+                        builder[referencedProjectId] = reverseDependencies.Remove(projectId);
+                    }
+                }
+            }
+
+            // Finally, remove 'projectId' itself.
+            builder.Remove(projectId);
+            return builder.ToImmutable();
+        }
+
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForRemovedProjectReference(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId,
+            ProjectId referencedProjectId)
+        {
+            if (existingReverseReferencesMap.TryGetValue(referencedProjectId, out var referencingProjects))
+            {
+                return existingReverseReferencesMap.SetItem(referencedProjectId, referencingProjects.Remove(projectId));
+            }
+
+            return existingReverseReferencesMap;
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_referencesMap"/> for the removal of a project.
+        /// </summary>
+        /// <param name="existingReverseReferencesMap">The <see cref="_lazyReverseReferencesMap"/> prior to the removal.
+        /// This map serves as a hint to the removal process; i.e. it is assumed correct if it contains data, but may be
+        /// omitted without impacting correctness.</param>
+        /// <param name="referencesMap">The <see cref="_referencesMap"/> prior to the removal.</param>
+        /// <param name="projectId">The ID of the project which is being removed.</param>
+        /// <returns>The <see cref="_referencesMap"/> for the project dependency graph once the project is removed.</returns>
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
```

Comment:
See other comment: surprised this is nullable. The null case is also just the empty case...?

Reply:
This is only empty if `_projectIds` is empty.

:arrow_discussion: Moving discussion to https://github.com/dotnet/roslyn/pull/40358#discussion_r371570675